### PR TITLE
[interfaces] add modeling from robynpy_interfaces

### PR DESCRIPTION
**Pull request by alxlyj**
## Project Robyn

**Summary:** We are including the interfaces (methods and data classes) for the modeling subcomponent in Python. This interface is inferred from model.R.  

## What is new:
- feat: New Python interfaces for modeling subcomponent. 

### Class diagram
```mermaid
classDiagram
    class BaseModelExecutor {
        <<abstract>>
        +MMMData mmmdata
        +HolidaysData holidays_data
        +Hyperparameters hyperparameters
        +CalibrationInput calibration_input
        +FeaturizedMMMData featurized_mmm_data
        +model_run(dt_hyper_fixed: Dict, ts_validation: bool, add_penalty_factor: bool, refresh: bool, seed: int, cores: int, trials_config: TrialsConfig, rssd_zero_penalty: bool, objective_weights: Dict, nevergrad_algo: NevergradAlgorithm, intercept: bool, intercept_sign: str, outputs: bool, model_name: Models)* ModelOutputs
    }

    class ModelExecutor {
        +model_run(dt_hyper_fixed: Dict, ts_validation: bool, add_penalty_factor: bool, refresh: bool, seed: int, cores: int, trials_config: TrialsConfig, rssd_zero_penalty: bool, objective_weights: Dict, nevergrad_algo: NevergradAlgorithm, intercept: bool, intercept_sign: str, outputs: bool, model_name: Models) ModelOutputs
    }

    class RidgeModelBuilder {
        +MMMData mmm_data
        +HolidaysData holiday_data
        +CalibrationInput calibration_input
        +Hyperparameters hyperparameters
        +FeaturizedMMMData featurized_mmm_data
        +build_models(trials_config: TrialsConfig, dt_hyper_fixed: DataFrame, ts_validation: bool, add_penalty_factor: bool, seed: int, rssd_zero_penalty: bool, objective_weights: List, nevergrad_algo: NevergradAlgorithm, intercept: bool, intercept_sign: str, adstock: AdstockType) ModelOutputs
        -_model_train(hyper_collect: Hyperparameters, trials_config: TrialsConfig, cores: int, intercept_sign: str, intercept: bool, nevergrad_algo: NevergradAlgorithm, dt_hyper_fixed: DataFrame, ts_validation: bool, add_penalty_factor: bool, objective_weights: Dict, rssd_zero_penalty: bool, seed: int) ModelOutputs
        +run_nevergrad_optimization(hyper_collect: Hyperparameters, iterations: int, cores: int, nevergrad_algo: NevergradAlgorithm, intercept: bool, intercept_sign: str, ts_validation: bool, add_penalty_factor: bool, objective_weights: Dict, dt_hyper_fixed: DataFrame, rssd_zero_penalty: bool, trial: int, seed: int) List[Trial]
        +model_decomp(coefs: Dict, y_pred: ndarray, dt_mod_saturated: DataFrame, dt_saturated_immediate: DataFrame, dt_saturated_carryover: DataFrame, dt_mod_roll_wind: DataFrame, refresh_added_start: str)* ModelDecompOutput
        -_model_refit(x_train: ndarray, y_train: ndarray, x_val: ndarray, y_val: ndarray, x_test: ndarray, y_test: ndarray, lambda_: float, lower_limits: List, upper_limits: List, intercept: bool, intercept_sign: str)* ModelRefitOutput
        -_lambda_seq(x: ndarray, y: ndarray, seq_len: int, lambda_min_ratio: float)* ndarray
        -_hyper_collector(adstock: str, all_media: List, paid_media_spends: List, organic_vars: List, prophet_vars: List, context_vars: List, dt_mod: DataFrame, hyper_in: Dict, ts_validation: bool, add_penalty_factor: bool, dt_hyper_fixed: DataFrame, cores: int)* HyperCollectorOutput
    }

    class MMMData {
        +DataFrame data
        +MMMDataSpec mmmdata_spec
        +display_data() None
        +get_summary() DataFrame
        +add_column(column_name: str, data: List) None
        +remove_column(column_name: str) None
    }

    class HolidaysData {
        +DataFrame dt_holidays
        +List[ProphetVariableType] prophet_vars
        +List[ProphetSigns] prophet_signs
        +str prophet_country
    }

    class Hyperparameters {
        +Dict[str, ChannelHyperparameters] hyperparameters
        +AdstockType adstock
        +float lambda_
        +List[float] train_size
        +get_hyperparameter(channel: str) ChannelHyperparameters
        +has_channel(channel: str) bool
        +get_hyperparameter_limits()* Dict[str, List[str]]
    }

    class CalibrationInput {
        +Dict[str, ChannelCalibrationData] channel_data
    }

    class FeaturizedMMMData {
        +DataFrame dt_mod
        +DataFrame dt_modRollWind
        +Dict modNLS
    }

    class ModelOutputs {
        +List[Trial] trials
        +str train_timestamp
        +int cores
        +int iterations
        +bool intercept
        +str intercept_sign
        +str nevergrad_algo
        +bool ts_validation
        +bool add_penalty_factor
        +Dict hyper_updated
        +bool hyper_fixed
        +Dict convergence
        +Any ts_validation_plot
        +str selectID
        +int seed
    }

    class Trial {
        +DataFrame resultHypParam
        +DataFrame xDecompAgg
        +DataFrame liftCalibration
        +DataFrame decompSpendDist
        +float nrmse
        +float decomp_rssd
        +float mape
    }

    class TrialsConfig {
        +int trials
        +int iterations
        +int cores
    }

    class NevergradAlgorithm {
        <<enumeration>>
        DE
        TWO_POINTS_DE
        ONE_PLUS_ONE
        DOUBLE_FAST_GA_DISCRETE_ONE_PLUS_ONE
        DISCRETE_ONE_PLUS_ONE
        PORTFOLIO_DISCRETE_ONE_PLUS_ONE
        NAIVE_TBPSA
        CGA
        RANDOM_SEARCH
    }

    class Models {
        <<enumeration>>
        RIDGE
    }

    BaseModelExecutor <|-- ModelExecutor
    ModelExecutor --> RidgeModelBuilder : uses
    BaseModelExecutor --> MMMData : uses
    BaseModelExecutor --> HolidaysData : uses
    BaseModelExecutor --> Hyperparameters : uses
    BaseModelExecutor --> CalibrationInput : uses
    BaseModelExecutor --> FeaturizedMMMData : uses
    RidgeModelBuilder --> ModelOutputs : produces
    RidgeModelBuilder --> Trial : produces
    ModelOutputs --> Trial : contains
    RidgeModelBuilder --> TrialsConfig : uses
    RidgeModelBuilder --> NevergradAlgorithm : uses
    ModelExecutor --> Models : uses
```

### Parity:
```mermaid
graph TB

    subgraph R["R Implementation (model.R)"]
        R1[robyn_run]
        R2[robyn_train]
        R3[robyn_mmm]
        R4[model_decomp]
        R5[model_refit]
        R6[lambda_seq]
        R7[hyper_collector]
        R8[init_msgs_run]
    end

    subgraph Python["Python Implementation"]
        P1[ModelExecutor.model_run]
        P2[RidgeModelBuilder.build_models]
        P3[RidgeModelBuilder._model_train]
        P4[RidgeModelBuilder.run_nevergrad_optimization]
        P5[RidgeModelBuilder.model_decomp]
        P6[RidgeModelBuilder._model_refit]
        P7[RidgeModelBuilder._lambda_seq]
        P8[RidgeModelBuilder._hyper_collector]
        P9[init_msgs_run]
    end

    %% Connections
    R1 ---- P1
    R2 ---- P2
    R3 ---- P3
    R3 ---- P4
    R4 ---- P5
    R5 ---- P6
    R6 ---- P7
    R7 ---- P8
    R8 ---- P9

    %% New classes and data structures
    P10[BaseModelExecutor]
    P11[TrialsConfig]
    P12[FeaturizedMMMData]
    P13[ModelOutputs]
    P14[Trial]
    P15[NevergradAlgorithm]

    %% Connections to new classes
    P1 --> P10
    P1 --> P11
    P1 --> P12
    P1 --> P13
    P13 --> P14
    P1 --> P15
    P2 --> P11
    P2 --> P15

    %% Legend
    classDef implemented fill:#8FBC8F,stroke:#2F4F4F;  % Darker green with darker stroke
    classDef new fill:#4682B4,stroke:#191970;         % Steel blue with navy stroke

    class R1,R2,R3,R4,R5,R6,R7,R8,P1,P2,P3,P4,P5,P6,P7,P8,P9 implemented;
    class P10,P11,P12,P13,P14,P15 new;

    legendA[Implemented]
    legendB[New Python Classes/Structures]

    class legendA implemented;
    class legendB new;
```

-  No implementations have been included just yet with this PR so it would be lightweight to land. More rigorous testing would be included in the implementation PR.

### CODE REVIEW

File: python/src/robyn/modeling/base_model_executor.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        featurized_mmm_data: FeaturizedMMMData,
+    ) -> None:
+        self.mmmdata = mmmdata
+        self.holidays_data = holidays_data
+        self.hyperparameters = hyperparameters
+        self.calibration_input = calibration_input
+        self.featurized_mmm_data = featurized_mmm_data
+
+    @abstractmethod
+    def model_run(
```
Comment (sumane81) [inline python/src/robyn/modeling/base_model_executor.py:35]: @gufengzhou can you verify these parameters for 1. any duplicates with instance variables 2. any logical grouping that can be done for these parameters.

File: python/src/robyn/modeling/entities/enums.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+from enum import Enum, auto
+
+
+class NevergradAlgorithm(Enum):
```
Comment (sumane81) [inline python/src/robyn/modeling/entities/enums.py:4]: @gufengzhou do we support all of the algorithms listed below? We will keep list of supported algorithms only here.

File: python/src/robyn/modeling/entities/enums.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+
+class Models(Enum):
+    """
+    Enumeration of available model types.
+
+    This enum can be expanded to include other model types as they are implemented.
+    """
+
+    RIDGE = auto()  # Ridge Regression
```
Comment (sumane81) [inline python/src/robyn/modeling/entities/enums.py:29]: @gufengzhou  Any other models we support?

File: python/src/robyn/modeling/entities/modeloutputs.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    xDecompAgg: pd.DataFrame
+    liftCalibration: pd.DataFrame
+    decompSpendDist: pd.DataFrame
+    nrmse: float
+    decomp_rssd: float
+    mape: float
+
+
+@dataclass
+class ModelOutputs:
```
Comment (sumane81) [inline python/src/robyn/modeling/entities/modeloutputs.py:31]: @gufengzhou Please verify if any attributes or information missing from model run output.

File: python/src/robyn/modeling/entities/modelrun_trials_config.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+            determines how many times the optimization algorithm will attempt
+            to improve the model within each trial.
+        cores (Optional[int]): The number of CPU cores to use for parallel
+            processing. If None, the system will determine the optimal number
+            of cores to use based on available resources.
+    """
+
+    trials: int
+    iterations: int
+    cores: Optional[int] = None
```
Comment (sumane81) [inline python/src/robyn/modeling/entities/modelrun_trials_config.py:25]: @alxlyj Let's remove cores from trials config and keep it independent variable. It doesn't really below to trials config

**Review by sumane81 — CHANGES_REQUESTED** [review]
@gufengzhou requested verification of method parameters and logical grouping wherever it is done.
@alxlyj minor change requested to remove cores from trials config.

File: python/src/robyn/modeling/entities/modelrun_trials_config.py (hunk 5 above)
Comment (alxlyj) [reply python/src/robyn/modeling/entities/modelrun_trials_config.py]: removed cores from trials config data class.

File: python/src/robyn/modeling/entities/modeloutputs.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        resultHypParam (pd.DataFrame): Hyperparameters and their values for this trial.
+        xDecompAgg (pd.DataFrame): Aggregated decomposition results for this trial.
+        liftCalibration (pd.DataFrame): Lift calibration results for this trial.
+        decompSpendDist (pd.DataFrame): Decomposition spend distribution for this trial.
+        nrmse (float): Normalized Root Mean Square Error for this trial.
+        decomp_rssd (float): Decomposition Root Sum Squared Distance for this trial.
+        mape (float): Mean Absolute Percentage Error for this trial.
+    """
+
+    resultHypParam: pd.DataFrame
```
Comment (sumane81) [inline python/src/robyn/modeling/entities/modeloutputs.py:21]: nit; Update variable names to snake_case

File: python/src/robyn/modeling/entities/enums.py (hunk 2 above)
Comment (sumane81) [reply python/src/robyn/modeling/entities/enums.py:4]: Seems TWO_POINTS_DE is default but it can take other algorithms.

**Review by sumane81 — APPROVED** [review]

**Pull request merged**